File: java/5.java
class Solution {
    public String longestPalindrome(String s) {
        String result = "";
        for (int start = 0; start < s.length(); start++) {
            String oddString = checkResursively(s, start, start);
            if (oddString.length() > result.length()) {
                result = oddString;
            }
            int end = start + 1;
            if (end < s.length() && s.charAt(start) == s.charAt(end)) {
                String evenString = checkResursively(s, start, end);
                if (evenString.length() > result.length()) {
                    result = evenString;
                }
            }
        }
        return result;
    }

    public String checkResursively(String s, int start, int end) {
        int newStart = start - 1;
        int newEnd = end + 1;
        if (newStart < 0 || newEnd == s.length()) {
            return s.substring(start, end + 1);
        } else {
            if (s.charAt(newStart) == s.charAt(newEnd)) {
                return checkResursively(s, newStart, newEnd);
            } else {
                return s.substring(start, end + 1);
            }
        }
    }
}

File: python/5.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if s == "":
            return ""
        dp = [[0] * len(s) for _ in range(len(s))]
        longest_str = s[0]
        for i in range(len(s)):
            dp[i][i] = 1
        for i in range(len(s) - 1):
            if s[i] == s[i + 1]:
                dp[i][i + 1] = 1
                if (len(longest_str) < len(s[i : i + 2])):
                    longest_str = s[i : i + 2]
        for diff in range(2, len(s)):
            for i in range(0, len(s) - diff):
                j = i + diff
                if s[i] == s[j]:
                    dp[i][j] = dp[i + 1][j - 1]
                if dp[i][j] == 1:
                    if (len(longest_str) < len(s[i : j + 1])):
                        longest_str = s[i : j + 1]
        return longest_str
